### PR TITLE
Docs: manually adding external packages in modules

### LINKED ISSUE #25612: External cray-mpich provides incorrect spec['mpi'].prefix.include path
### Steps to reproduce

This is observed on the spock system, a cray early-access machine at ORNL. It causes several mpi dependent package builds to fail including TAU, which I will use as an example. 

packages.yaml looks like:

```
packages:
    mpich:
      buildable: false
      externals:
      - spec: mpich@8.1.7
        modules:
        - cray-mpich/8.1.7
```

The TAU package sets up its mpi configuration like this:
```
            options.append("-mpiinc=%s" % spec['mpi'].prefix.include)
            options.append("-mpilib=%s" % spec['mpi'].prefix.lib)
```

Which translates to this in the actual configure invocation:
`'-mpiinc=/opt/cray/pe/mpich/8.1.7/gtl/include' '-mpilib=/opt/cray/pe/mpich/8.1.7/gtl/lib'`

However, /opt/cray/pe/mpich/8.1.7/gtl/include does not exist. The include files for this mpi install are located in compiler specific subdiriectories of /opt/cray/pe/mpich/8.1.7/ofi and /opt/cray/pe/mpich/8.1.7/ucx. (e.g. /opt/cray/pe/mpich/8.1.7/ofi/gnu/9.1/include)

This makes the system-provided mpich impossible to use as an external, at least for packages that use spec['mpi'].prefix.include to get the include path.

### Error message

_No response_

### Information on your system

* **Spack:** 0.16.2-4019-80713e234c
* **Python:** 3.6.13
* **Platform:** cray-sles15-zen2
* **Concretizer:** original

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by christoph-conrads**
~~fixes #25612~~

After having trouble to use MPI on two supercomputers with Spack (see #12520), I decided to look at the Spack source  code. The `module show` output is critically important to the use of external packages in modules. This pull request adds a subsection explaining how to handle
* modules with dependencies on other modules,
* modules where the prefix computation does not work, and
* metamodules (i.e., modules that only load other modules).

I want to thank the Spack team for the clean code base and spelling out function names. Within hours I had a rough grasp of how the prefix is computed inside the Spack code. (It could have been faster if I had read the [Developer Guide](https://spack.readthedocs.io/en/latest/developer_guide.html) but adding `print()` statements was more engaging.)

### CODE REVIEW

**Comment by christoph-conrads** [comment]
```
 Codecov / codecov/project failed 2 minutes ago in 0s
80.15% (-2.74%) compared to d27e0bf
```
I guess I did not find enough readers.

**Comment by christoph-conrads** [comment]
~~Fixed #25612 with the advice in the new documentation~~

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
This LGTM, added other reviewers too since I am not working regularly with externals AND modules so I may be missing some fine points.

**Comment by christoph-conrads** [comment]
The pull request was updated to highlight the importance of picking the correct package name based on the [comment by lukebroskop](https://github.com/spack/spack/issues/25612#issuecomment-915741227) and the [comment by haampie](https://github.com/spack/spack/issues/25612#issuecomment-915972874) regarding the use of `mpich` and `cray-mpich`.

**Comment by christoph-conrads** [comment]
Hi @scheibelp, could you review this PR, please? The PR was originally created in response to #12520 that you assigned yourself.

File: lib/spack/docs/build_settings.rst [hunk 1]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+Manually Adding External Packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section discusses how to add external packages to the ``packages.yaml``
+file when modules depend on other modules or when Spack is unable to compute
+the prefix. When using modules, Spack will attempt to parse the ``module show``
+output in order to determine all relevant settings and variables. Reasons why
+this may not work automatically are
+
+* the wrong package name,
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: As in: the module name doesn't match the package name?

File: lib/spack/docs/build_settings.rst [hunk 2]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+* missing module dependencies,
+* nonstandard paths, and
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: By this do you mean: The modules are not available in `MODULEPATH`?

File: lib/spack/docs/build_settings.rst [hunk 3]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+    - gcc/8.3.1 cuda/10.1.2
+    - gcc/8.3.1 cuda/10.1.1
+
+    If you want to use this module with another software environment,
+    please contact the support team.
+    -------------------------------------------------------------------
+
+There are two things of importance to note. First, there are multiple possible
+module combinations to satisfy the compiler and CUDA dependency; for this
+example will use ``gcc/8.3.1`` and ``cuda/10.1.2``. Second, the output does not
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: "will" -> "we will"

You can say

"This example will use..."

or

"For this example we will use..."

But not

"For this example will use..."

File: lib/spack/docs/build_settings.rst [hunk 4]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+
+    conflict	 mpi
+    prepend-path	 PATH /usr/lib64/openmpi/bin
+    prepend-path	 LD_LIBRARY_PATH /usr/lib64/openmpi/lib
+    prepend-path	 PYTHONPATH /usr/lib64/python2.7/site-packages/openmpi
+    prepend-path	 MANPATH /usr/share/man/openmpi-x86_64
+    [snip]
+
+Given this module output, ``spack external find`` will determine
+``/usr/lib64/openmpi`` as the prefix when it is actually ``/usr``. This
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: Are there also headers in this case? Where are they stored? If it's something like `/usr/include/openmpi` then FWIW there is no prefix that Spack can use generally: we expect that all the package's files are contained in a single prefix for now, e.g. `/usr/{lib, include}` or `/path/to/openmpi/{lib, include}`. There might be per-package support for this in a couple cases, but generally an external which behaves this way will require symlinked directories to create the illusion of a unified prefix.

File: lib/spack/docs/build_settings.rst [hunk 5]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+output in order to determine all relevant settings and variables. Reasons why
+this may not work automatically are
+
+* the wrong package name,
+* missing module dependencies,
+* nonstandard paths, and
+* the use of metamodules, i.e., modules whose only purpose is to load other
+  modules.
+
+For certain libraries there exist multiple implementations (e.g., MPI). If you
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: This paragraph feels out of place: it has no prior introduction. Should it actually be a 5th bullet point (in addition to the 4 above)? A separate section?

File: lib/spack/docs/build_settings.rst [hunk 6]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+modify the ``packages.yaml`` file, then you must make sure to select the
+correct package name. For example, the package ``mpich`` can be used only for
+the vanilla MPICH implementation but not for Cray MPICH; Cray MPICH has its own
+package called ``cray-mpich``. If the wrong module name is picked, this can
+cause errors later, e.g., Spack may compute a wrong prefix.
+
+The case of missing dependencies will be discussed based on the real-world
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst:248]: I think it would be useful to add minor section headers for each of the cases you describe initially (e.g. have a section titled "Missing dependencies").

File: lib/spack/docs/build_settings.rst [hunk 7]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+show openmpi/4.0.2`` output.
+
+.. code-block:: console
+
+    $ module show openmpi/4.0.2-cuda
+    -------------------------------------------------------------------
+    /gpfslocalsup/pub/modules-idris-env4/modulefiles/linux-rhel8-skylake_avx512/openmpi/4.0.2-cuda:
+
+    module-whatis   {An open source Message Passing Interface implementation.}
+    prereq          intel-compilers/19.0.4 pgi/20.1 pgi/19.10 gcc/10.1.0 gcc/8.3.1
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst:259]: As a side note: we don't currently search through modules at all to detect new packages (e.g. `module avail`), as mentioned above (in this file) at https://spack.readthedocs.io/en/latest/build_settings.html?highlight=external#automatically-find-external-packages. If we add that, we should probably consider checking for prereq modules and trying to load them first.

File: lib/spack/docs/build_settings.rst [hunk 8]
```diff
@@ -209,6 +209,170 @@ Specific limitations include:
...
+
+.. _manually-adding-external-packages:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Manually Adding External Packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section discusses how to add external packages to the ``packages.yaml``
+file when modules depend on other modules or when Spack is unable to compute
+the prefix. When using modules, Spack will attempt to parse the ``module show``
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: Overall I think this should be phrased differently (but it would take me significant time to come up with suggestions): our support for finding externals via modules is nonexistent (mentioned above in https://spack.readthedocs.io/en/latest/build_settings.html?highlight=external#automatically-find-external-packages). So I think this should emphasize that and overall that explain that this section describes how to resolve some of the resulting shortcomings.

As it is, this introduction names specific issues with using module-based externals so I think is too focused.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Overall I like the content but have some issues with the organization and introduction (and in a few spots ask for clarification). I also have a question about the openmpi/CentOS7 use case: we might not be able to handle that well.

File: lib/spack/docs/build_settings.rst (hunk 1 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: As in: the package name does not match the software loaded by the module. The text was modified but what I committed a minute ago is still misleading.

File: lib/spack/docs/build_settings.rst (hunk 2 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: No, the directory structure of the software loaded by the module does not match conventions, e.g., headers in `/prefix/include` and the library in `/prefix/lib{,32,64}`. Text modified.

File: lib/spack/docs/build_settings.rst (hunk 3 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: Fixed, thank you for the suggestions.

File: lib/spack/docs/build_settings.rst (hunk 5 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: Minor headers were added as proposed below and a sentence saying that each bullet point will be addressed one by one.

File: lib/spack/docs/build_settings.rst (hunk 6 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst:248]: Added

File: lib/spack/docs/build_settings.rst (hunk 4 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: > Where are they stored?

On CentOS 7 with OpenMPI, in `/usr/include/openmpi-x86_64`.

> There might be per-package support for this in a couple cases, but generally an external which behaves this way will require symlinked directories to create the illusion of a unified prefix.

Yes, this is really a mess but at least for OpenMPI on CentOS, the world adapted to it (e.g., the CMake `FindMPI` module finds OpenMPI on CentOS).

File: lib/spack/docs/build_settings.rst (hunk 8 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: New wording:
```asciidoc
So far this section described external package detection and its limitations.
The paragraphs below deal with challenges arising when external packages are
found in loadable modules. When using modules, Spack will attempt to parse the
``module show`` output in order to determine all relevant settings and
variables. Reasons why this may not work automatically are
```

> As it is, this introduction names specific issues with using module-based externals so I think is too focused.

The problem for me is that without this section, I cannot get MPI to run (see #12520) and without MPI, Spack becomes hard to sell.

Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: Thank you for the review and your improvement proposals. Please comment on my recent changes.

File: lib/spack/docs/build_settings.rst [hunk 9]
```diff
@@ -209,6 +209,192 @@ Specific limitations include:
...
+
+The case of missing dependencies will be discussed based on the real-world
+example of loading OpenMPI 4.0.2 with CUDA support. Shown below is the ``module
+show openmpi/4.0.2`` output.
+
+.. code-block:: console
+
+    $ module show openmpi/4.0.2-cuda
+    -------------------------------------------------------------------
+    /gpfslocalsup/pub/modules-idris-env4/modulefiles/linux-rhel8-skylake_avx512/openmpi/4.0.2-cuda:
```
Comment (adamjstewart) [inline lib/spack/docs/build_settings.rst]: I would like to avoid user-specific paths in the docs. Can we use `/usr/local` or something like that instead?

**Comment by haampie** [comment]
Failing CI with

> OSError: [Errno 2] No such file or directory: ‘/tmp/tmp4kqIAZ/S.gpg-agent.extra’

should be fixed with #26665

File: lib/spack/docs/build_settings.rst (hunk 9 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: All of the outputs were created using available modules on real clusters (here: Jean-Zay ran by the **I**nstitut du **d**éveloppement et des **r**essources en **i**nformatique **s**cientifique) but this can be replaced. Do you want me to use `/usr/local` as a generic prefix everywhere?

Comment (adamjstewart) [reply lib/spack/docs/build_settings.rst]: Yeah let's replace it

Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: * replaced `/gpfslocalsup/pub/modules-idris-env4/modulefiles/linux-rhel8-skylake_avx512/openmpi/4.0.2-cuda` with `/usr/local/modulefiles/openmpi/4.0.2-cuda`
* replaced `/gpfslocalsup/spack_soft/openmpi/4.0.2/gcc-8.3.1-n6vcsair26tkpepojy3c2gqxtqccijq3` with `/usr/local/spack_soft/openmpi/4.0.2/gcc-8.3.1-n6vcsair26tkpepojy3c2gqxtqccijq3`

**Comment by scheibelp** [comment]
@spackbot re-run pipeline

File: lib/spack/docs/build_settings.rst [hunk 10]
```diff
@@ -209,6 +209,192 @@ Specific limitations include:
...
+
+Each following paragraph will discuss one of the bullet points above.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Module/Package Name Mismatches
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For certain libraries there exist multiple implementations (e.g., MPI). If you
+modify the ``packages.yaml`` file, then you must make sure to select the
+correct package name. For example, the package ``mpich`` can be used only for
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: I would replace

> If you modify the ``packages.yaml`` file, then you must make sure to select the correct package name

with

> When defining an external entry for a package, make sure the package name identifies the implementation that matches the module.

But otherwise this paragraph looks good.

File: lib/spack/docs/build_settings.rst [hunk 11]
```diff
@@ -209,6 +209,192 @@ Specific limitations include:
...
+and all of its dependencies in the ``packages.yaml`` file to arrive at the
+following OpenMPI entry:
+
+.. code-block:: yaml
+
+    packages:
+      openmpi:
+        externals:
+        - spec: openmpi@4.0.2+cuda+cxx+cxx_exceptions~java~memchecker+pmi~sqlite3+static~thread_multiple~wrapper-rpath
+            fabrics=psm2 schedulers=slurm
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst:315]: Note this example could also record the `gcc` version used by the external version in the spec:

```
- spec: openmpi@4.0.2 %gcc@8.3.1
      +cuda+cxx+cxx_exceptions~java~memchecker+pmi~sqlite3
      +static~thread_multiple~wrapper-rpath fabrics=psm2 schedulers=slurm
```

Currently it's not possible to record the `^cuda@10.1.2` bit (although on Cray machines at least theres some work to ingest full DAGs into the DB: https://github.com/spack/spack/pull/24894).

File: lib/spack/docs/build_settings.rst (hunk 4 above)
Comment (scheibelp) [reply lib/spack/docs/build_settings.rst]: > ``spack external find`` will determine ``/usr/lib64/openmpi`` as the prefix when it is actually ``/usr``.

After thinking about this I find it confusing to phrase it this way:

* It's true that using a lower-level prefix like `/usr/` successfully finds the libs and headers, but keep in mind this requires a recursive file search which can be time-consuming
* In Spack terms, `/usr` is also not the right prefix because e.g. openmpi doesnt store an `openmpi` exe in `/usr/bin`. Since this loads the module, that doesn't matter. But my point is that it's more direct to phrase this in terms of finding a prefix which helps locate the libs/headers/binaries rather than characterizing a prefix as right or wrong
* Note that if we updated the implementation of `headers` in the `openmpi` package, we could check for `os=centos` and search for headers in `/usr/include/openmpi-x86_64` if that's the case. I think I would prefer reorienting the section toward this approach

File: lib/spack/docs/build_settings.rst [hunk 12]
```diff
@@ -209,6 +209,192 @@ Specific limitations include:
...
+    module          load intel-mkl/19.0.4
+    module          load intel-mpi/19.0.4
+    module          load intel-vtune/19.0.4
+    module          load intel-advisor/19.0.4
+    module          load intel-tbb/19.0.4
+    module          load intel-itac/19.0.4
+    -------------------------------------------------------------------
+
+The module ``intel-all`` is obviously a metamodule because it only loads other
+modules and its ``module show`` output cannot be used to derive the needed
```
Comment (scheibelp) [inline lib/spack/docs/build_settings.rst]: I would rephrase:

> and its ``module show`` output cannot be used to derive the needed environment changes to use

as

> ; the module contents do not include direct manipulations of `PATH`, which Spack depends on to determine the package prefix.

Also regarding: 

> For this reason, metamodules *cannot* be used in ``packages.yaml`` files.

Could a metamodule be used *in combination* with a lower level module? e.g. if the metamodule loads needed dependencies it might be faster than the example in the section "Missing Modules Dependencies"

I might rephrase that line as

> For this reason, metamodules cannot generally be used in place of the direct package module when defining external packages.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Apologies this took a while to get back to. I have a few more suggestions. I think they are generally straightforward except my comment on the section "Nonstandard Directory Structure" (but let me know if any of them are confusing or disagreeable).

**Comment by scheibelp** [comment]
@christoph-conrads do you think you'd be able to address the latest review requests in the next week or so?

File: lib/spack/docs/build_settings.rst (hunk 11 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst:315]: This is verbatim Spack output. Shall I add the compiler dependency anyway?

File: lib/spack/docs/build_settings.rst (hunk 12 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: fixed in c766605dcf825b9062b426799e22bf63cb384808

File: lib/spack/docs/build_settings.rst (hunk 10 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: replacement added in c766605dcf825b9062b426799e22bf63cb384808

File: lib/spack/docs/build_settings.rst (hunk 4 above)
Comment (christoph-conrads) [reply lib/spack/docs/build_settings.rst]: New proposed wording:
```asciidoc
Given this module output, ``spack external find`` will determine
``/usr/lib64/openmpi`` as the prefix when only ``/usr`` allows Spack to
determine all required paths.
```

**Comment by alecbcs** [comment]
Closing due to no response. @christoph-conrads happy to reopen if you'd like to continue working on this PR.

**Comment by christoph-conrads** [comment]
There was a response (see my force push right above the latest posts) but I should not have relied on GitHub notifying the reviewer. Since I am not working in HPC anymore, I will not reopen the PR.
